text/3371-cargo-target-dir-templates.md

```diff
@@ -0,0 +1,423 @@
+
+To prevent collisions by craftings paths, the `<manifest-path-hash>` directory will be computed from a hash of the workspace manifest's full path (and possibly other data, for example `bazel` uses its version and the current user too).
+
+### Symbolic links
+
+In the following situation
+
+```
+/Users/
+├─ poliorcetics/
+│  ├─ projects/
+│  │  ├─ actual-crate/
+│  │  │  ├─ Cargo.toml
+│  │  ├─ symlink-to-crate/ -> actual-crate/
+```
+
+When calling `cargo metadata` in the `symlink-to-crate` path, the result contains `"manifest_path": "/Users/poliorcetics/projects/actual-crate/Cargo.toml"` and `"workspace_root":"/Users/poliorcetics/projects/actual-crate"`. This behaviour means that symlinks won't change the final directory used inside `{manifest-path-hash}`, or in other words: symbolic links are resolved.
+
+### Handle possibly thousands of directories in a single templated `CARGO_TARGET_DIR` path
+
+While a single dev machine is unlikely to have enough projects that the naming scheme of `<manifest-path-hash>` will produce enough directories to slow down working in `$CARGO_TARGET_DIR/`, it could still happen, and notably in private CI, which are often less compartimentalized than public ones. Simple cruft over time (i.e, never calling `cargo clean` over years) could also make it happen, if much slower.
+
+To prevent this, `cargo` splits the hash into something like `$CARGO_TARGET_DIR/hash[:2]/hash[2:4]/hash[4:]/...`. Since the naming scheme is considered an implementation detail, if this prove insufficient it could be changed in a subsequent version of `cargo`.
+
+## Impact on `cargo ...` calls
+
+When calling `cargo` with a builtin call (e.g., `build`, `check` or `test`) where a templated `CARGO_TARGET_DIR` is active, `cargo` will first resolve the effective `CARGO_TARGET_DIR` and then proceed with the command as if `CARGO_TARGET_DIR` had been set directly. For third party tools (`cargo-*`), where cargo does not know about the relevant `Cargo.toml`, the tool will have to use [`cargo_metadata`](https://docs.rs/cargo_metadata), as is already expected today, to learn about the effective target directory.
+
+In the same vein, `cargo metadata` fills the target directory information with the absolute path and make no mention of the template in `CARGO_TARGET_DIR` since it can only be used with a single workspace at once.
+
+### `cargo clean`
+
+Currently, if `CARGO_TARGET_DIR` is set to anything but `target` for a project, `cargo clean` does not delete the `target/` directory if it exists, instead deleting the directory pointed by `CARGO_TARGET_DIR`. The same behavior is used for the templated version: if it set, `cargo clean` deletes `/path/to/<manifest-path-hash>/` and not `target/`.
+
+### Providing backlinks
```

Comment:
If I'm understanding correctly, this RFC specifies no backlink behavior.  If so, this doesn't belong in the Reference section.  Maybe as an Alternative?  We have the GC part of this already in the Future Possibilities.

Reply:
Done in https://github.com/rust-lang/rfcs/pull/3371/commits/11ab0e38c6351d8d89fa2a6abb063f19c2a18916